keps/sig-network/4969-cluster-domain-pod-api/README.md

```diff
@@ -0,0 +1,957 @@
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add a new field to the Pod status, containing the cluster domain suffix.
+It should also be accessible via the downward API and `EnvVarSource.fieldRef`.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+The Pod `foo` needs to access its sibling Service `bar` in the same namespace.
+It adds two `env` bindings:
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: foo
+spec:
+  containers:
+    - name: foo
+      env:
+        - name: NAMESPACE
+          valueFrom:
+            fieldRef: metadata.namespace
+        - name: CLUSTER_DOMAIN
+          valueFrom:
+            fieldRef: status.dns.clusterDomain
+```
+
+`foo` can now perform the query by running `curl http://bar.$NAMESPACE.svc.$CLUSTER_DOMAIN/`.
+
+(Of course, in practice this would likely be integrated into the app itself, not by shelling into bash, but the principle still applies.)
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Split Domains
+
+Kubernetes currently does not prohibit different kubelets from specifying 
+different cluster domains (`node-a` could set `cluster.local` while `node-b` 
+specifies `cluster.remote`).
+Advertising FQDNs generated using this API could cause issues in these mixed
+environments, since `node-b` might not be able to resolve `cluster.local`
+FQDNs correctly.
+
+## Design Details
+
+A new field `dns.clusterDomain` would be added to `PodStatus`:
+
+```go
+type PodStatus struct {
+    // ..existing fields elided
+    Dns PodDNSStatus
+}
+
+type PodDNSStatus struct {
+    ClusterDomain string
+}
+```
+
+The field would be populated by the Kubelet during pod initialization.
```

Comment:
edge case, imagine a cluster admin decides to change the cluster domain, so it changes the domain the kubelet config and restarts the kubelet , do we:
1. update the existing pod status to reflect the current cluster domain
2. leave old pods with the old clusterdomain

Reply:
Status should reflect what the pod was started with (which is also in /etc/resolv.conf)

Reply:
We can't make it literally immutable, since kubelet might update it, but it's a reflection of truth, not a source of truth. :)

Reply:
ok, then this is good as it is